vpr/src/place/noc_place_utils.h

```diff
@@ -0,0 +1,375 @@
+ * indicate that the traffic flow uses them.
+ * 
+ * @param traffic_flow_id The traffic flow to re-route.
+ * @param noc_traffic_flows_storage Contains all the traffic flow information
+ * within the NoC. Used to get the current traffic flow information.
+ * @param noc_model Contains all the links and rotuters within the NoC. Used
+ * to route traffic flows within the NoC.
+ * @param noc_flows_router The packet routing algorithm used to route traffic
+ * flows within the NoC.
+ * @param placed_cluster_block_locations A datastructure that identifies the
+ * placed grid locations of all cluster blocks.
+ */
+void re_route_traffic_flow(NocTrafficFlowId traffic_flow_id, NocTrafficFlows& noc_traffic_flows_storage, NocStorage& noc_model, NocRouting* noc_flows_router, const vtr::vector_map<ClusterBlockId, t_block_loc>& placed_cluster_block_locations);
+
+/**
+ * @brief 
+ * 
+ * @param new_noc_aggregate_bandwidth_cost 
+ * @param new_noc_latency_cost 
+ */
+void recompute_noc_costs(double* new_noc_aggregate_bandwidth_cost, double* new_noc_latency_cost);
+
+/**
+ * @brief Updates all the cost normalization factors relevant to the NoC.
+ * Also updates the placement cost depending the on the placment mode.
+ * Handles exceptional cases so that the normalization factors do not
+ * reach INF.
+ * This is intended to be used to initialize the normalization factors of
+ * the NoC and also at the outer loop iteration of placement to 
+ * balance the NoC costs with other placment cost parameters.
+ * 
+ * @param costs Contains the normalization factors which need to be updated
+ * @param placer_opts Determines the placement mode
+ */
+void update_noc_normalization_factors(t_placer_costs& costs, const t_placer_opts& placer_opts);
+
+/**
+ * @brief Calculates the aggregate bandwidth of each traffic flow in the NoC
+ * and initializes local variables that keep track of the traffic flow 
+ * aggregate bandwidths cost.
+ * Then the total aggregate bandwidth cost is determines by summing up all
+ * the individual traffic flow aggregate bandwidths.
+ * 
+ * This should be used after initial placement to determine the starting
+ * aggregate bandwdith cost of the NoC.
+ * 
+ * @return double The aggregate bandwidth cost of the NoC.
+ */
+double comp_noc_aggregate_bandwidth_cost(void);
+
+/**
+ * @brief Calculates the latency cost of each traffic flow in the NoC
+ * and initializes local variables that keep track of the traffic flow
+ * latency costs. Then the total latency cost is determined by summing up all
+ * the individual traffic flow latency costs.
+ * 
+ * This should be used after initial placement to determine the starting latency
+ * cost of the NoC.
+ * 
+ * @return double The latency cost of the NoC.
+ */
+double comp_noc_latency_cost(const t_noc_opts& noc_opts);
+
+/**
+ * @brief 
+ * 
+ */
+int check_noc_placement_costs(const t_placer_costs& costs, double error_tolerance, const t_noc_opts& noc_opts);
```

Comment:
empty comment

Reply:
Mention this is a brute-force recompute from scratch (including determining all the traffic routes).

Reply:
Updated comment